src/rpc/tungstenite_client/client.rs

```diff
@@ -0,0 +1,330 @@
+	pub fn send_extrinsic(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<<SubmitOnlyHandler as HandleMessage>::ThreadMessage>,
+	) -> Result<()> {
+		self.start_rpc_client_thread(json_req, result_in, SubmitOnlyHandler::default())
+	}
+
+	pub fn send_extrinsic_until_ready(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<<SubmitAndWatchHandler as HandleMessage>::ThreadMessage>,
+	) -> Result<()> {
+		self.start_rpc_client_thread(
+			json_req,
+			result_in,
+			SubmitAndWatchHandler::new(XtStatus::Ready),
+		)
+	}
+
+	pub fn send_extrinsic_and_wait_until_broadcast(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<<SubmitAndWatchHandler as HandleMessage>::ThreadMessage>,
+	) -> Result<()> {
+		self.start_rpc_client_thread(
+			json_req,
+			result_in,
+			SubmitAndWatchHandler::new(XtStatus::Broadcast),
+		)
+	}
+
+	pub fn send_extrinsic_and_wait_until_in_block(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<<SubmitAndWatchHandler as HandleMessage>::ThreadMessage>,
+	) -> Result<()> {
+		self.start_rpc_client_thread(
+			json_req,
+			result_in,
+			SubmitAndWatchHandler::new(XtStatus::InBlock),
+		)
+	}
+
+	pub fn send_extrinsic_and_wait_until_finalized(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<<SubmitAndWatchHandler as HandleMessage>::ThreadMessage>,
+	) -> Result<()> {
+		self.start_rpc_client_thread(
+			json_req,
+			result_in,
+			SubmitAndWatchHandler::new(XtStatus::Finalized),
+		)
+	}
+
+	pub fn start_rpc_client_thread<MessageHandler>(
+		&self,
+		json_req: String,
+		result_in: ThreadOut<MessageHandler::ThreadMessage>,
+		message_handler: MessageHandler,
+	) -> Result<()>
+	where
+		MessageHandler: HandleMessage<Error = (RpcClientError, bool), Context = MySocket>
+			+ Clone
+			+ Send
+			+ 'static,
+		MessageHandler::ThreadMessage: Send + Sync + Debug,
+		MessageHandler::ThreadMessage: From<MessageHandler::Result>,
+	{
+		let url = Url::parse(self.url.as_str()).map_err(|e| RpcClientError::Other(e.into()))?;
+		let max_attempts = self.max_attempts;
+
+		thread::spawn(move || {
+			let mut current_attempt: u8 = 1;
+			let mut socket: MySocket;
+
+			while current_attempt <= max_attempts {
+				match connect_with_config(url.clone(), None, max_attempts) {
+					Ok(res) => {
+						socket = res.0;
+						let response = res.1;
+						debug!(
+							"Connected to the server. Response HTTP code: {}",
+							response.status()
+						);
+						match socket.write_message(Message::Text(json_req.clone())) {
+							Ok(_) => {},
+							Err(e) => {
+								error!("write msg error:{:?}", e);
+							},
+						}
+						if socket.can_read() {
+							current_attempt = 1;
+							// After sending the subscription request, there will be a response(result)
+							let msg_from_req = read_until_text_message(&mut socket);
+							match msg_from_req {
+								Ok(msg_from_req) => {
+									debug!("response message: {:?}", msg_from_req);
+									loop {
+										let msg = read_until_text_message(&mut socket);
```

Comment:
That's a loop within a loop. I'd rather avoid this.

How about getting rid of the loops with one `while` looking something like this:
```rust
while Ok(message) = socket.read_message() {
     match message {
         Message::Text(s) => { // do something .. } ,
         Message::Close(close) => { break; },  // Don't think we should continue the loop if a close message is received
         _ => { debug!("ignore non text message"); },
     }
} 
```
That will make it much more obvious on what's happening within the loop.

Reply:
I also think that's a lot of dubplicate code. Maybe you could extract a function and reuse it in the message handlers.